CRAMcodecs.tex

```diff
@@ -0,0 +1,3181 @@
+      \State $R_j \gets$\ \Call{RansRenorm}{$R_j$}
+      \State $L_j \gets s$
+    \EndFor
+    \State $i \gets i+1$
+  \EndWhile
+  \Statex \ \ \ \ (Now deal with the remainder if buffer size is not a multiple of 4,)
+  \Statex \ \ \ \ (using rANS state 3 exclusively.)
+  \For{$i \gets i \times 4 \algorithmicto len-1$}
+    \State $f \gets$ \Call{RansGetCumulativeFreq}{$R_3$}
+    \State $s \gets$ \Call{RansGetSymbolFromFreq}{$C_{L_3},\ f$}
+    \State $output_i \gets s$
+    \State $R_3 \gets$\ \Call{RansAdvanceStep}{$R_3,\ C_{L_3,s},\ F_{L_3,s}$}
+    \State $R_3 \gets$\ \Call{RansRenorm}{$R_3$}
+    \State $L_3 \gets s$
+  \EndFor
+\EndProcedure
+\end{algorithmic}
+
+\section{rANS Nx16}
+
+CRAM version 3.1 defines an additional rANS entropy encoder, using
+16-bit renormalisation instead of the 8-bit used in CRAM 3.0 and with
+inbuilt bit-packing and run-length encoding.  The Order-1 rANS Nx16
+encoder has also been modified to use a maximum frequency of 1024
+instead of 4096.  This offers better cache performance for the large
+Order-1 tables and usually has minimal impact on compression ratio.
+(The 4x8 rANS entropy encoder is still available if the reduced
+frequency range is a problem.)
+
+Additionally it permits adjustment of the number of interleaved rANS
+states from the fixed 4 used in rANS 4x8 to either 4 or 32 states.
+The benefit of the 32-way interleaving is in enabling efficient use of
+SIMD instructions for faster encoding and decoding speeds.  However it
+has a small cost in size and initialisation times so it is not
+recommended on smaller blocks of data.
+
+Frequencies are now stored using uint7 format instead of ITF8.  The
+tables are also stored differently, separating the list of symbols
+present in the alphabet (those with frequency greater than zero) from
+the frequencies themselves.  For the Order-1 frequency table this list
+of symbols is those used in any context, thus we only have one
+alphabet recorded for all contexts.  This means in some contexts
+some (potentially many) symbols will have zero frequency.  To reduce
+the Order-1 table size an additional zero run-length encoding step is
+used.  Finally the Order-1 frequencies may optionally be compressed
+using the Order-0 rANS Nx16 codec.
+
+Frequencies must always add up to a power of 2, but do not necessarily
+have to match the final power of two used in the Order-0 (4096) and
+Order-1 (1024, 4096) entropy decoder algorithm.  A normalisation step is
+applied after reading the frequencies to scale them appropriately.
+This is required as the Order-1 frequencies may be scaled differently
+for each context.
+
+\subsection{Frequency tables}
+
+\begin{algorithmic}[1]
+\Statex (Reads a list of symbols $A$ used in our alphabet)
+\Function{ReadAlphabet}{}
+\State $s \gets$ \Call{ReadUint8}{}
+\State $last\_sym \gets s$
+\State $rle \gets 0$
+\Repeat
+  \State $A \gets (A,s)$ \Comment{Append $s$ to the symbol list $A$}
```

Comment:
This currently implies that `A` holds a list of symbols in that order that they are read from the file, which has implications for the order of frequency values read later.  It might be useful to add a restriction that `s` is always increasing, or maybe that `A` should either be sorted into order before returning it, or should be treated as an ordered set rather than a list.

See also https://github.com/samtools/htscodecs/pull/26/files#r814844095

Reply:
The fact this is a run-length encoded list of symbols with the ability to auto-increment symbol numbers is at leas very heavily implying that they are stored sequentially, however I agree it needs to be an explicit statement as technically it could be encoded as a mix of out of order symbols interspersed with ordered clumps of adjacent values.  I'll amend.